chore(protocol): refactor some protocol implementations

lbwtw opened the pull request.

zccrs committed:
WIP

zccrs committed:
WIP

zzxyb committed:
refactor gamma woutput cursorshape fractionalscale protocol imp

Groveer committed:
fix(wserver): no match function for call std::forward (#2)

zccrs commented on examples/tinywl/helper.h:
```diff
@@ -72,6 +76,10 @@ class Helper : public WSeatEventFilter {
     Q_INVOKABLE void onSurfaceLeaveOutput(WToplevelSurface *surface, WSurfaceItem *surfaceItem, WOutput *output);
     std::pair<WOutput*,OutputInfo*> getFirstOutputOfSurface(WToplevelSurface *surface);
 
+    QWFractionalScaleManagerV1 *fractionalScaleManagerV1() const;
```
有必要加成员函数暴漏出去吗

zccrs commented on examples/tinywl/main.cpp:
```diff
@@ -110,6 +116,18 @@ void Helper::initProtocols(WServer *server, WOutputRenderWindow *window, QQmlEng
+    m_gammaControlManager = server->attach<WGammaControlManagerV1>();
+    connect(m_gammaControlManager, &WGammaControlManagerV1::gammaChanged, this, [this]
+            (WOutput *output, wlr_gamma_control_v1 *gamma_control,
+             size_t ramp_size, uint16_t *r, uint16_t *g, uint16_t *b){
```
){ 之间加个空格

zccrs commented on src/server/kernel/woutput.cpp:
```diff
@@ -6,7 +6,7 @@
 #include "wseat.h"
 #include "wtools.h"
-#include "wquickgammacontrol_p.h"
+#include "wgammacontrolmanagerv1.h"
```
没用到？是不是可以直接删掉

zccrs commented on src/server/protocols/wgammacontrolmanagerv1.cpp:
```diff
@@ -0,0 +1,99 @@
+{
+    W_D(WGammaControlManagerV1);
+
+    return wlr_gamma_control_manager_v1_get_control(d->nativeHandle(), output);
```
这不应该是在 qlwroots 里封装吗？

zccrs commented on src/server/protocols/wgammacontrolmanagerv1.cpp:
```diff
@@ -0,0 +1,99 @@
+    W_D(WGammaControlManagerV1);
+
+    m_handle = QWGammaControlManagerV1::create(server->handle());
+    QObject::connect(handle(), &QWGammaControlManagerV1::gammaChanged, this, [this]
```
这就没必要封装了

zccrs commented on src/server/protocols/wcursorshapemanagerv1.cpp:
```diff
@@ -106,27 +115,40 @@ static inline auto wpToWCursorShape(wp_cursor_shape_device_v1_shape shape) {
+
+    if (!m_handle) {
+        m_handle = QWCursorShapeManagerV1::create(server->handle(), CURSOR_SHAPE_MANAGER_V1_VERSION);
+        QObject::connect(handle(), &QWCursorShapeManagerV1::requestSetShape, this, [this]
```
没啥封装的必要性，在 tinywl 里自己收原始信号处理

lbwtw replied:
这个需要封装

justforlxz committed:
Move WQuickForeignToplevelManagerV1 to protocols/WForeignToplevel (#4)

Dami-star committed:
move WQuickXdgOutputManager to protocols/WXdgOutput (#10)

Changed from inheriting from WQuickWaylandServerInterface to inheriting
from WServerInterface.
